hack/generator/pkg/codegen/pipeline_collapse_cross_group_refs.go

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package codegen
+
+import (
+	"context"
+
+	"github.com/Azure/azure-service-operator/hack/generator/pkg/astmodel"
+	"github.com/pkg/errors"
+)
+
+// collapseCrossGroupReferences finds and removes references between API groups. This isn't particularly common
+// but does occur in a few instances, for example from Microsoft.Compute -> Microsoft.Compute.Extensions.
+func collapseCrossGroupReferences() PipelineStage {
+	return MakePipelineStage(
+		"collapseCrossGroupReferences",
+		"Finds and removes cross group references",
+		func(ctx context.Context, types astmodel.Types) (astmodel.Types, error) {
+			resources := astmodel.CollectResourceDefinitions(types)
+			result := make(astmodel.Types)
+
+			for resourceName := range resources {
+				walker := newTypeWalker(types, resourceName)
+				updatedTypes, err := walker.Walk(types[resourceName])
+				if err != nil {
+					return nil, errors.Wrapf(err, "failed walking types")
+				}
+
+				for _, newDef := range updatedTypes {
+					err := result.AddAllowDuplicates(newDef)
+					if err != nil {
+						return nil, err
+					}
+				}
```

Comment:
I think you could replace this loop with the new `OverlayWith()` method.

Reply:
I'm not finding such a method?

Reply:
SEe PR #1598

Reply:
I went to do this, but realized that `OverlayWith` doesn't have the `AllowDuplicates` behavior. I don't think we want to just pick a winner in the case of a collision (which is what `OverlayWith` does) as that is making an implicit assumption that the loser has the same structure as the winner. The loop as written ensures that in cases where there is a collision, the two types are identical and thus can be treated the same.

Reply:
Fair enough.